### ISSUE TITLE
Support hard line breaks

### ISSUE THREAD

**Issue by jordanbtucker** [charmbracelet/glow#212]
Please add support for hard line breaks.

https://spec.commonmark.org/0.30/#hard-line-breaks

**Comment by tricarte** [comment]
Yes, please.

**Comment by gennaro-tedesco** [comment]
I too endorse this.

**Comment by wantyapps** [comment]
Wait... isn't markdown newline equal to two enters?

**Comment by jordanbtucker** [comment]
@gennaro-tedesco Your code *should not* work. In order to have a single line break be interpreted as a new line, you need to precede it with [at least two spaces](https://spec.commonmark.org/0.29/#hard-line-breaks), like this (select the text to see that there are two spaces after `hello`):

```md
hello  
world
```

Alternatively, you can end your line with a backslash `\` like this:

```md
hello\
world
```

Neither of the above examples work in glow though, which is why I created this issue.

@wantyapps

> Wait... isn't markdown newline equal to two enters?

Not quite. Two line breaks in Markdown result in a new `<p>` element whereas a line break preceded by either two or more spaces or a `\` results in a `<br>` element.

Compare [Paragraphs](https://spec.commonmark.org/0.29/#paragraphs) with [Hard line breaks](https://spec.commonmark.org/0.29/#hard-line-breaks).

**Comment by gennaro-tedesco** [comment]
Thank you for the clarification @jordanbtucker, I did not know, so bad on my side; however I do confirm that neither of the above (proper) examples work in Glow.

**Comment by muesli** [comment]
You're right, we're reflowing the text, but I think we can make that optional fairly easily.

**Comment by wantyapps** [comment]
> @gennaro-tedesco Your code *should not* work. In order to have a single line break be interpreted as a new line, you need to precede it with [at least two spaces](https://spec.commonmark.org/0.29/#hard-line-breaks), like this (select the text to see that there are two spaces after `hello`):
> 
> 
> 
> ```md
> 
> hello  
> 
> world
> 
> ```
> 
> 
> 
> Alternatively, you can end your line with a backslash `\` like this:
> 
> 
> 
> ```md
> 
> hello\
> 
> world
> 
> ```
> 
> 
> 
> Neither of the above examples work in glow though, which is why I created this issue.
> 
> 
> 
> @wantyapps
> 
> 
> 
> > Wait... isn't markdown newline equal to two enters?
> 
> 
> 
> Not quite. Two line breaks in Markdown result in a new `<p>` element whereas a line break preceded by either two or more spaces or a `\` results in a `<br>` element.
> 
> 
> 
> Compare [Paragraphs](https://spec.commonmark.org/0.29/#paragraphs) with [Hard line breaks](https://spec.commonmark.org/0.29/#hard-line-breaks).

Thank you for the clarification.

**Comment by seanlaguna** [comment]
This is pretty crucial for me, I'm really enjoying glow for reading my markdown notes in terminal, but I use double-space line breaks all over the place and right now there's no way to get a single line break to render via glow at all as far as I can tell!

**Comment by magnus-ISU** [comment]
Is there a reason this has not been implemented yet? This makes glow render markdown incorrectly, in a pretty serious way

**Comment by fiskhest** [comment]
Looks like this is solved in the upstream dependency https://github.com/charmbracelet/glamour/pull/115, I updated glamour version in `go.mod`:
`github.com/charmbracelet/glamour v0.3.1-0.20210906181425-86a99924d7f6`

and added:
```
diff --git a/main.go b/main.go
index b36d902..1852704 100644
--- a/main.go
+++ b/main.go
@@ -248,6 +248,7 @@ func executeArg(cmd *cobra.Command, arg string, w io.Writer) error {
                gs,
                glamour.WithWordWrap(int(width)),
                glamour.WithBaseURL(baseURL),
+               glamour.WithPreservedNewLines(),
        )
        if err != nil {
                return err
```

and my example file `test.md` renders correctly:
```
hello  
world
```

```
$ glow test.md
hello
world
```

@muesli @meowgorithm Any chance we could see a new release for glow addressing this issue? Looks simple enough, though I'm probably overlooking something (I'm new to go/the project)

**Comment by muesli** [comment]
@fiskhest Working on it! Just not sure yet if we shouldn't keep the current setting as our default.

**Comment by fiskhest** [comment]
That's great, thank you!

I might be misunderstanding your statement, but the current setting is not adhering to the markdown specification and as such should not be the default. If anything, the opposite should be the default, and the current behavior require setting a param.

**Comment by magnus-ISU** [comment]
I agree that the current behaviour shouldn't be preserved, I don't know why you would even offer a compatability flag with the current output, as it isn't standard and I really doubt there is anyone who *wanted* the behaviour of two-spaces-ending-a-line to not be a line break.

Also, does it also work with backslashes ending a line rather than 2 spaces?

**Comment by fiskhest** [comment]
Any updates on this matter @muesli ?

**Comment by jjc224** [comment]
Also curious about updates on this? @muesli 
I agree that this behavior should be the default.
This issue unfortunately makes using Glow unusable for me. It's a pretty significant compatibility issue.

**Comment by Antiz96** [comment]
I think I have the same behaviour.
A "<br>" tag is shown as a double space rather than a break line in glow.

```
$ cat test.md
Hello
<br>
World
```
Expectation:
```
$ glow test.md
  Hello
  World
```
Reality:
```
$ glow test.md

  Hello  World

```
I don't know if this is directly linked to this issue as you guys seem to talk about double space and not specifically about "<br>" tags but that's the same thing right ? (Sorry if I'm wrong).

Anyway, I don't know if it is still planned to fix this (as it has been raised a long time ago) but I would also be interested by a fix :)
I wish I could help more with that...

By the way, thanks for your amazing work !

**Comment by jordanbtucker** [comment]
@Antiz96 That is a separate but valid issue. I recommend opening an issue for it in the [glamour repo](https://github.com/charmbracelet/glamour), which is the code base that performs the markdown conversion. Technically, this issue should be there too.

**Comment by Antiz96** [comment]
@jordanbtucker Thanks for your answer !
I didn't knew about glamour. After looking at the currently opened issues in the glamour repo, it seems like a similar issue has already been raised [here](https://github.com/charmbracelet/glamour/issues/90). 
It is mostly about double spaces and back slashes, not specifically about `<br>` tags tho... 
But if at least one of those 3 options works, I'd be fine with it :)

I'm still going to open an issue anyway in order to get it documented.

Thanks again !

**Comment by fiskhest** [comment]
@jordanbtucker technically this issue is solved in glamour, only need a default setting flipped in glow.

**Comment by Antiz96** [comment]
@fiskhest @jordanbtucker 
So should I still post in glamour repo or glow then ?

**Comment by fiskhest** [comment]
@Antiz96 sadly, I don't know or have the time to verify that your issue is resolved in glamour or not - you could follow the instructions posted in [my comment]( https://github.com/charmbracelet/glow/issues/212#issuecomment-921632355) earlier in this issue thread and verify that yourself.

**Comment by WieeRd** [comment]
@muesli Are there any future plans for this?
I think this is a serious problem making glow unusable and inconsistent with other md viewers.  
It should be specified in the README along with the plans and current workaround.

**Comment by krader1961** [comment]
@muesli, FWIW, this impacts the `help` command I'm implementing for the Elvish shell. There are a few commands with explicitly multiline usage text. For example: https://elv.sh/ref/builtin.html#str-cmp. That is rendered thusly by `glow` regardless of whether I end each line with two spaces, a backslash or `<br/>`:

```
<  $number... # less  <= $number... # less or equal  == $number... # equal  !=
$number... # not equal  >  $number... # greater  >= $number... # greater or
equal
```

**Comment by WieeRd** [comment]
I've tested @fiskhest's solution (https://github.com/charmbracelet/glow/issues/212#issuecomment-921632355) and it still works very well.  
As you can see from above I've created a PR #362 based on that.  
Hopefully this is the right way to fix it and gets merged soon.

**Comment by jordanbtucker** [comment]
> Working on it! Just not sure yet if we shouldn't keep the current setting as our default.

This would technically be a breaking change. So, if users are relying on current behavior, or if they expect minor versions of glow to not change the output, then I would recommend taking a [Semantic Versioning](https://semver.org/) approach and implement CommonMark compliant output behind an optional flag in this version. When you release v2, the setting can be turned on by default.

If you've already been changing output in minor versions of glow (i.e. not adhering to semver) then I would just turn the setting on by default.

**Comment by magnus-ISU** [comment]
Please don't make this not default. People either don't use this functionality or came to complain about it on this issue. Just my thoughts.

**Comment by fiskhest** [comment]
> Please don't make this not default. People either don't use this functionality or came to complain about it on this issue. Just my thoughts.

I second this motion.

**Comment by jordanbtucker** [comment]
I've updated my comments to point to the latest version of the [CommonMark spec](https://spec.commonmark.org/0.30/). No conformance changes were made to hard line breaks. Only changes to the prose occurred in that section.

**Comment by jordanbtucker** [comment]
@krader1961 Wouldn't that be better placed in a code block, like it is in the [doc comment](https://github.com/elves/elvish/blob/b47f8e0dc217f7eedc855e69f621d1a259c430e8/pkg/eval/builtin_fn_str.go#L20-L27)? Glow will maintain line breaks inside code blocks, even without two spaces at the end of the line.

**Comment by WieeRd** [comment]
I think preserving current behavior shouldn't even be an option.  
I wrote some thoughts about it in https://github.com/charmbracelet/glow/pull/362#issuecomment-1188688606.  
TLDR it should be applied right away in the next release as default.

**Comment by jordanbtucker** [comment]
@magnus-ISU I wasn't saying that they should change their Markdown to accommodate glow. I was saying that the [example](https://elv.sh/ref/builtin.html#str-cmp) they referenced is already in a code block, so I don't see why they would need to use hard line breaks at all.

But you do make a good point that Markdown users generally don't write Markdown to render with glow, at least not exclusively. They write it to render as HTML typically on some platform like GitHub or in a static site generator like Jekyll, and rendering a line break by ending a line with two spaces has been a feature of Markdown since its [inception](https://web.archive.org/web/20040607100324/https://daringfireball.net/projects/markdown/syntax#p). So, regardless of breaking changes, the current behavior should be treated like a bug, and I don't see a reason to keep the behavior either.

**Comment by WieeRd** [comment]
Rest In Peace, #212.  
2020.12.01 ~ 2023.01.24  
You'll not be missed.

**Comment by jordanbtucker** [comment]
One step closer to resolving https://github.com/charmbracelet/glamour/issues/122

**Comment by geekstakulus** [comment]
I am a little confused about this issue, because I am not ending a line with two spaces or with \, but line breaks are still rendered in my output. Is there any setting that I can change to modify this behaviour? I work with semantic line feeds and I only want line breaks when I instruct it to by following the standard. A single line break shouldn't render a line break, or have I interpreted this wrong?

Thanks in advance!

**Comment by acvgo** [comment]
Using 1.5.1 - Glow is not respecting two-space endings.
Example file and screen cap attached. 

[README.md](https://github.com/charmbracelet/glow/files/15277462/README.md)
<img width="811" alt="2024-05-10 11_23_10-" src="https://github.com/charmbracelet/glow/assets/105814448/88a48858-401f-41d1-98ba-5091eb56f709">
